### PR TITLE
Softmax

### PR DESCRIPTION
**Pull request by apaszke**
- Cleaned up THNN and THCUNN code and kernels
- Improved THCUNN kernel performance 5x, making it match cuDNN performance
- Added support for computing softmax over arbitrary dims.
- Deprecated implicit dim selection in softmax/softmin/log_softmax
- Both functions now accept inputs with arbitrarily many dimensions
- Autograd functions no longer save the input (it's unnecessary)
- Added cuDNN bindings for softmax, but they are unused as THCUNN
  matches or even exceeds cuDNN performance

cc: @karpathy

### CODE REVIEW

File: torch/csrc/cudnn/Descriptors.h [hunk 1]
```diff
@@ -4,27 +4,67 @@
...
+  TensorDescriptor(TensorDescriptor&& ref) { desc = ref.desc; ref.desc = NULL; }
   ~TensorDescriptor() {
     cudnnDestroyTensorDescriptor(desc);
   }
   void set(cudnnDataType_t dataType, int dim, int* size, int* stride) {
     CHECK(cudnnSetTensorNdDescriptor(desc, dataType, dim, size, stride));
   }
+  void set(const at::Tensor &t, int pad = 0) {
+    int dim = t.ndimension();
+    if (dim > 10 || pad > 10)
```
Comment (ngimel) [inline torch/csrc/cudnn/Descriptors.h]: CUDNN_DIM_MAX is 8, defined in cudnn.h. Better to use it rather than string literal.

**Comment by colesbury** [comment]
The changes to `gen_variable_type.py` look good.

I don't really like the idea of adding a bunch of cuDNN binding code that isn't actually used anywhere. That increases technical debt without enough benefit.

**Comment by colesbury** [comment]
I've written this before, but I also think that using `dim=1` as the default is a mistake, since it doesn't generalize well to n-dimensions and doesn't work on 1-D arrays. I don't think we should be baking in assumptions about data layout where we don't need to (i.e. that 3D is temporal with `batch x channel x time`).

**Comment by soumith** [comment]
@colesbury isn't it over-generalization to avoid the dominant use-cases?

**Comment by jekbradbury** [comment]
This is awesome, but I would personally prefer in the absence of consensus between defaulting to dim 1 and dim -1 (I lean towards -1) that the defaults stay as they were in previous versions to reduce the amount of code that has to change. Anything with configurable dimension is great, though, and updating code isn't too hard.

**Comment by soumith** [comment]
okay, adam can you just keep the defaults as-is (i think it's stupid, but whatever).
can you do that and rebase the PR, i want to merge this soon.

**Comment by apaszke** [comment]
I'll fix it tomorrow

**Comment by colesbury** [comment]
I don't want to delay this any further, so if you don't like this suggestion just ignore it. 

Another option is deprecate the call without a dim. This keeps backwards compatibility, but let's us eventually move away from the old behavior. e.g.

```python
def softmax(input, dim=None):
  if dim is None:
    dim = ... # old behavior
    # raise deprecated warning
  return _Softmax(dim)(input)
```

**Comment by apaszke** [comment]
Yeah I think deprecating implicit dim choosing for now is reasonable. It'll give us more time to decide on default, but I really think we should do something with what we have right now

**Comment by jekbradbury** [comment]
I see a new LogSoftMax kernel in THCUNN in the diff, but [not](https://github.com/pytorch/pytorch/blob/89bae46f2b100e418fed71e1694cc83d8d74c333/torch/lib/THCUNN/generic/SoftMax.cu) a new SoftMax kernel. Or is this still WIP?

**Comment by apaszke** [comment]
It's not WIP. I didn't have time to fix plain SoftMax. The existing kernels are quite complicated, I'll dig into them later

File: torch/lib/ATen/gen.py [hunk 2]
```diff
@@ -218,6 +218,7 @@ def generate_storage_type_and_tensor(backend, density, scalar_type, declarations
 
 cwrap_files = [f for f in files if f.endswith('.cwrap')]
 nn_files = [f for f in files if f.endswith('.yaml') or f.endswith('.h')]
+print(nn_files)
```
Comment (colesbury) [inline torch/lib/ATen/gen.py]: remove please

File: torch/lib/ATen/nn_parse.py [hunk 3]
```diff
@@ -145,6 +145,8 @@ def arg_expr(prefix, suffix):
         return {'type': 'EXPRESSION', 'name': expr}
 
     thnn_args = []
+    print(thnn_function.arguments)
```
Comment (colesbury) [inline torch/lib/ATen/nn_parse.py]: remove please

File: torch/lib/ATen/nn_parse.py [hunk 4]
```diff
@@ -220,6 +222,9 @@ def base_declaration(func, thnn_function, backends):
     """Creates the NN function without any buffers in it's signature"""
     name, params = re.match(NAME_PARAM_REGEX, func['name']).groups()
     params = params.split(', ')
+    print('-' * 80)
```
Comment (colesbury) [inline torch/lib/ATen/nn_parse.py]: remove please

**Review by colesbury — APPROVED** [review]
lgtm - please remove the debug print() statements (and fix the lint)

**Comment by colesbury** [comment]
@apaszke, is this ready to merge?

**Comment by apaszke** [comment]
Yes, except that I just realized that I forgot to remove cuDNN bindings

**Comment by apaszke** [comment]
This is ready to merge (py2.7 failure is a flaky VolumetricMaxPooling test). Can I merge this with an ATen change, or do I have to send a separate PR there?

**Comment by apaszke** [comment]
Will fail until zdevito/ATen#106 is merged to PyTorch
